File: test_dir/python/src/numpy_arrays/exceptions.py
import numpy as np

from test_dir.python.pypp.libs.numpy.creation import pypp_np_zeros, pypp_np_array
from test_dir.python.pypp.libs.numpy.impl import PyppNpArr


def numpy_array_exceptions_fn():
    print("NUMPY ARRAY EXCEPTIONS RESULTS:")
    a: PyppNpArr[np.float32] = pypp_np_zeros([2, 3], np.float32)
    try:
        a(0, 0, 0)
    except IndexError as e:
        print("index error: " + str(e))
    try:
        a(99, 0)
    except IndexError as e:
        print("index error: " + str(e))
    try:
        a(0, 98)
    except IndexError as e:
        print("index error: " + str(e))
    empty_list: list[int] = []
    try:
        pypp_np_zeros(empty_list, np.float32)
    except ValueError as e:
        print("value error: " + str(e))
    try:
        pypp_np_array(empty_list, np.float32)
    except ValueError as e:
        print("value error: " + str(e))
    try:
        pypp_np_zeros([-1, 2], np.float32)
    except ValueError as e:
        print("value error: " + str(e))
    try:
        pypp_np_array([[1, 2], [3, 4, 5]], np.float32)
    except ValueError as e:
        print("value error: " + str(e))

File: test_dir/cpp/src/numpy_arrays/exceptions.cpp
#include "numpy_arrays\exceptions.h"

void numpy_array_exceptions_fn() {
    print(PyStr("NUMPY ARRAY EXCEPTIONS RESULTS:"));
    NpArr<float> a = pypp_np::zeros<float>(PyList({2, 3}));
    try {
        a(0, 0, 0);
    } catch (const PyppIndexError &pypp_e) {
        std::string e = pypp_e.what();
        print(PyStr("index error: ") + to_pystr(e));
    }
    try {
        a(99, 0);
    } catch (const PyppIndexError &pypp_e) {
        std::string e = pypp_e.what();
        print(PyStr("index error: ") + to_pystr(e));
    }
    try {
        a(0, 98);
    } catch (const PyppIndexError &pypp_e) {
        std::string e = pypp_e.what();
        print(PyStr("index error: ") + to_pystr(e));
    }
    PyList<int> empty_list = PyList<int>({});
    try {
        pypp_np::zeros<float>(empty_list);
    } catch (const PyppValueError &pypp_e) {
        std::string e = pypp_e.what();
        print(PyStr("value error: ") + to_pystr(e));
    }
    try {
        pypp_np::array<float>(empty_list);
    } catch (const PyppValueError &pypp_e) {
        std::string e = pypp_e.what();
        print(PyStr("value error: ") + to_pystr(e));
    }
    try {
        pypp_np::zeros<float>(PyList({-1, 2}));
    } catch (const PyppValueError &pypp_e) {
        std::string e = pypp_e.what();
        print(PyStr("value error: ") + to_pystr(e));
    }
    try {
        pypp_np::array<float>(PyList({PyList({1, 2}), PyList({3, 4, 5})}));
    } catch (const PyppValueError &pypp_e) {
        std::string e = pypp_e.what();
        print(PyStr("value error: ") + to_pystr(e));
    }
}